src/protocol/sort/generate_permutation_opt.rs

```diff
@@ -113,6 +117,104 @@ where
+/// The function takes care of upgrading and validating while the sort protocol runs.
+/// It then returns a semi honest context with output in Replicated format. The caller should then upgrade the output and context before moving forward
+///
+/// Steps
+/// 1. [Malicious Special] Upgrade the context from semihonest to malicious and get a validator
+/// 2. [Malicious Special] Upgrade 0..`num_multi_bits` sort bit keys
+/// 3. Compute bit permutation that sorts 0..`num_multi_bits` bit
+///
+/// For `num_multi_bits` to N-1th bit of input share
+/// 1. i. Shuffle the i-1th composition
+///   ii. [Malicious Special] Validate the accumulator contents
+///  iii. [Malicious Special] Malicious reveal
+///   iv. [Malicious Special] Downgrade context to semihonest
+/// 2. i. [Malicious Special] Upgrade ith sort bit keys
+///   ii. Sort i..i+`num_multi_bits` bits based on i-1th bits by applying i-1th composition on i..i+`num_multi_bits` bits
+/// 3. Compute bit permutation that sorts i..i+`num_multi_bits` bits
+/// 4. Compute ith composition by composing i-1th composition on ith permutation
+/// In the end, following is returned
+///    i. n-1th composition: This is the permutation which sorts the inputs
+///   ii. Validator which can be used to validate the leftover items in the accumulator
+///
+/// ![Malicious sort permutation steps][malicious_sort]
+/// # Panics
+/// If sort keys dont have num of bits same as `num_bits`
+/// # Errors
+pub async fn malicious_generate_permutation_opt<'a, F>(
+    sh_ctx: SemiHonestContext<'a, F>,
+    sort_keys: &[Vec<Vec<Replicated<F>>>],
+) -> Result<(MaliciousValidator<'a, F>, Vec<MaliciousReplicated<F>>), Error>
+where
+    F: Field,
+{
+    let mut malicious_validator = MaliciousValidator::new(sh_ctx.narrow(&Sort(0)));
+    let mut m_ctx_bit = malicious_validator.context();
+    let input_len = u32::try_from(sort_keys[0].len()).unwrap(); // safe, we don't sort more than 1B rows
+
+    let upgraded_sort_keys = m_ctx_bit.upgrade(sort_keys[0].clone()).await?;
+    let lsb_permutation =
+        multi_bit_permutation(m_ctx_bit.narrow(&BitPermutationStep), &upgraded_sort_keys).await?;
+    let mut composed_less_significant_bits_permutation = lsb_permutation;
+
+    for (chunk_num, chunk) in sort_keys.iter().enumerate().skip(1) {
+        let revealed_and_random_permutations = malicious_shuffle_and_reveal_permutation(
+            m_ctx_bit.narrow(&ShuffleRevealPermutation),
+            input_len,
+            composed_less_significant_bits_permutation,
+            malicious_validator,
+        )
+        .await?;
+
+        malicious_validator =
+            MaliciousValidator::new(sh_ctx.narrow(&Sort(chunk_num.try_into().unwrap())));
+        m_ctx_bit = malicious_validator.context();
+        let upgraded_sort_keys = m_ctx_bit.upgrade(chunk.clone()).await?;
```

Comment:
For a follow-up PR:

I *think* it would be more efficient to wait to convert from XorReplicated sharings until *this point in time*. But let's revisit this later on.

Reply:
agreed, this should be doable

Reply:
Added a todo for this